# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1789418827 -7200
# Parent  a8f77a1b7029e8b288726927c91c8e516f901d45
NNN9 - Suppress unneeded logging to the error console.

The fix in ExtensionChild.sys.mjs solves "sendRemoveListener on closed conduit" - Bug 1633315.

diff --git a/toolkit/components/extensions/ExtensionChild.sys.mjs b/toolkit/components/extensions/ExtensionChild.sys.mjs
--- a/toolkit/components/extensions/ExtensionChild.sys.mjs
+++ b/toolkit/components/extensions/ExtensionChild.sys.mjs
@@ -709,8 +709,13 @@ export class ProxyAPIImplementation exte
     map.listeners.delete(listener);
     map.ids.delete(id);
     map.removedIds.add(id);
 
+    // The following call throws if there is no actor.
+    // No benefit in calling it.
+    if (!this.childApiManager.conduit.actor) {
+      return;
+    }
     this.childApiManager.conduit.sendRemoveListener({
       childId: this.childApiManager.id,
       listenerId: id,
       path: this.path,
diff --git a/toolkit/components/extensions/ExtensionParent.sys.mjs b/toolkit/components/extensions/ExtensionParent.sys.mjs
--- a/toolkit/components/extensions/ExtensionParent.sys.mjs
+++ b/toolkit/components/extensions/ExtensionParent.sys.mjs
@@ -331,9 +331,11 @@ const ProxyMessenger = {
 
   getTopBrowsingContextId(tabId) {
     // If a tab alredy has content scripts, no need to check private browsing.
     let tab = apiManager.global.tabTracker.getTab(tabId, null);
-    if (!tab || (tab.browser || tab).getAttribute("pending") === "true") {
+    // tab.getAttribute("pending") throws, so the following doesn't work:
+    // if (!tab || (tab.browser || tab).getAttribute("pending") === "true") {
+    if (!tab || !tab.browser || tab.browser.getAttribute("pending") === "true") {
       // No receivers in discarded tabs, so bail early to keep the browser lazy.
       throw new ExtensionError(ERROR_NO_RECEIVERS);
     }
     let browser = tab.linkedBrowser || tab.browser;
@@ -1254,9 +1256,17 @@ ParentAPIManager = {
     }
   },
 
   async recvAPICall(data, { actor }) {
-    let context = this.getContext(data.childId, actor);
+    let context;
+    // getContext will throw if the context has been destroyed
+    // in the meantime.
+    try {
+      context = this.getContext(data.childId, actor);
+    } catch (ex) {
+      console.info(`No context for ${data.childId}`);
+      return;
+    }
     let target = actor.browsingContext?.top.embedderElement;
 
     let reply = result => {
       if (target && !context.parentMessageManager) {
